### PR TITLE
Add support for python logging

### PR DESCRIPTION
**Pull request by cav71**
These are small changes in order to set/enable python logging.

It is heavily inspired by: https://github.com/abdnh/ankiutils/blob/master/src/ankiutils/log.py (@abdnh)

The reasons for this:
- given the logging is provided by the base python library it makes sense to leverage it
- use of python logging instead plain print statements
- no need for an external module (beside the anki main app)
- it makes easier to modify output streams in one place (the main app)
- format for the log output can be configured on a global/logger basis

Why not using the original code:
- logging should be part of the application/framework (eg. main anki package)
- propagation is good for dispatching records to the root logger (that prints to stdout by default)

TL;DR;
Client code (eg. addons) will call this to get a logging.Logger instance:
```
import logging
log = getattr(mw.addonManager, "getLogger", logging.getLogger)(__name__)
log.info("Hello from addon")
```

The message "Hello from addon" will be written to `pm.addonFolder()/NNNNNN/user_files/logs/NNNNNN.log` as well to stdout.

### CODE REVIEW

File: qt/aqt/addons.py [hunk 1]
```diff
@@ -665,6 +665,10 @@ def addonConfigHelp(self, module: str) -> str:
     def addonFromModule(self, module: str) -> str:
         return module.split(".")[0]
 
+    def getLogger(self, name):
+        from logging import getLogger
+        return getLogger(f"{aqt.log.LoggerManager.TAG}{self.addonFromModule(module)}")
```
Comment (abdnh) [inline qt/aqt/addons.py:670]: `module` is undefined. Also getLogger() is missing type hints.

File: qt/aqt/log.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    def __init__(self, path: Path | str, rootnode: logging.Logger):
+        super().__init__(rootnode)
+        self.path = Path(path)
+
+    # fixme: this seems not called
+    def _cleanup(self, handler: RotatingFileHandler):
+        from anki.hooks import wrap
+        from aqt.addons import AddonManager
+        AddonManager.deleteAddon = wrap(  # type: ignore[method-assign]
```
Comment (abdnh) [inline qt/aqt/log.py:40]: It's more clean/correct to call the cleanup method inside the relevant AddonManager methods instead of patching them.

File: qt/aqt/log.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    def getLogger(self, name: str) -> logging.Logger:
+        logger = super().getLogger(name)
+        if name.startswith(self.TAG) and RotatingFileHandler not in [handler.__class__ for handler in logger.handlers]:
+            path = self.path / f"{name.partition('.')[2]}.log"
+            path.parent.mkdir(parents=True, exist_ok=True)
+            handler = RotatingFileHandler(path)
+            logger.addHandler(handler)
+            try:
+                # Prevent errors when deleting/updating the add-on on Windows
```
Comment (abdnh) [inline qt/aqt/log.py:55]: Just a note that this is only required if the logs are saved inside the add-on's folder.

File: qt/aqt/addons.py (hunk 1 above)
Comment (cav71) [reply qt/aqt/addons.py:670]: fixed

File: qt/aqt/log.py (hunk 2 above)
Comment (abdnh) [reply qt/aqt/log.py:39]: Yes.

Comment (cav71) [reply qt/aqt/log.py]: Updated in [here](https://github.com/ankitects/anki/pull/2969/commits/62abefe33abed743f652fd8c4ba24f35c2a32188).

**Comment by dae** [comment]
What happens to existing add-ons that have registered their own logging if this code gets deployed?

I find myself wondering how necessary this is, but the authors of more in-depth add-ons may feel differently. @abdnh @glutanimate, do you have any thoughts?

**Comment by cav71** [comment]
> What happens to existing add-ons that have registered their own logging if this code gets deployed?

Not much, if they call it in the canonical way
```
log = logging.getLogger(__name__)
```
The output will bubble up to the root logger and printed to standard output (the default).

For a new addon they could call this:
```
log = mw.addonManager.getLogger(__name__)
```
Then the output will be logged under `pm.addonFolder()/NNNNNN/user_files/logs/NNNNNN.log` as well whatever the root logger emits to (by default stdout).

> I find myself wondering how necessary this is, but the authors of more in-depth add-ons may feel differently. @abdnh @glutanimate, do you have any thoughts?

The advantage is consolidating the logging emit mechanism into a single point, leaving client code to just report records. Potentially it can be extended to dump in a dialog in the main app or remote reporting with not much code change).

**Comment by abdnh** [comment]
> I find myself wondering how necessary this is, but the authors of more in-depth add-ons may feel differently. @abdnh @glutanimate, do you have any thoughts?

@dae I've not made much use of logging in add-ons yet. (The referenced code from my add-ons is a recent addition)
I'm also not sure how useful built-in logging will be. Personally, I'm not considering using this unless more logging-related changes are integrated into Anki in the future. I mean user-facing features such as a button to copy add-on logs in the add-ons dialog, for example.

> The advantage is consolidating the logging emit mechanism into a single point, leaving client code to just report records.

@cav71 The current setup emits logs with no timestamps, which is not very useful. Can we maybe set a default formatter?

File: qt/aqt/addons.py [hunk 4]
```diff
@@ -665,6 +670,10 @@ def addonConfigHelp(self, module: str) -> str:
     def addonFromModule(self, module: str) -> str:
         return module.split(".")[0]
 
+    def getLogger(self, module: str) -> logging.Logger:
```
[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/addons.py:674]: imported logging at module level, VScode should resolve them now

File: qt/aqt/addons.py [hunk 5]
```diff
@@ -729,6 +738,10 @@ def _userFilesBackupPath(self) -> str:
 
     def backupUserFiles(self, sid: str) -> None:
         p = self._userFilesPath(sid)
+
+        # close the aqt.log.RotatingFileHandler handler (and re-open)
+        aqt.log.close_module(module, reopen=True)
```
[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/addons.py:743]: fixed renaming sid -> module (it consistent with the rest of the code base).

File: qt/aqt/log.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        import aqt.log
+        aqt.log.config(
+            pm.addonFolder(),
+            level=logging.DEBUG
+        )
+    """
+    handlers = [
+        logging.StreamHandler(stream=sys.stdout),
+    ]
+    logging.basicConfig(handlers=handlers, **kwargs)
```
[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/log.py:67]: Would be sys.stderr better? Any suggestion?

**Comment by cav71** [comment]
> @cav71 The current setup emits logs with no timestamps, which is not very useful. Can we maybe set a default formatter?

The format is now nicer (and it can changed in a single line `FORMATTER = ...` in qt/aqt/log.py):
```
python qt/aqt/log.py
2024-01-29 23:01:07,910:INF:a: an info message
2024-01-29 23:01:07,910:WRN:a: a warning message
2024-01-29 23:01:07,911:INF:addon.X: an info message
2024-01-29 23:01:07,911:WRN:addon.X: a warning message
```
And
```
cat deleteme/X/user_files/logs/X.log
2024-01-29 23:01:07,911:INF:addon.X: an info message
2024-01-29 23:01:07,911:WRN:addon.X: a warning message
```

File: qt/aqt/log.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dae) [reply qt/aqt/log.py:96]: Logging to stderr would result in warning pop-ups.

This PR has introduced lines like '2024-01-30 15:37:44,937:WRN:waitress.queue: Task queue depth is 1
' in the console, which weren't previously shown. They aren't adding anything useful, and I'd prefer they weren't shown.

**Comment by dae** [comment]
> The advantage is consolidating the logging emit mechanism into a single point

But do we want that? Add-ons are authored by a disparate group of people, and each add-on developer may have their own preferences in terms of what and how things are logged, and at what level, if at all. We wouldn't even be consolidating the logs into one location.

File: qt/aqt/log.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/log.py:100]: I've silenced in the log.config: they are now gone.

**Comment by dae** [comment]
Thank you for elaborating. I'm going to defer to @abdnh and @glutanimate here - if one of them thinks merging this in is a good idea and the other doesn't object/want changes, I will be happy to do so once the tests pass.

**Review by glutanimate — COMMENTED** [review]
Thanks for your work on this! Left a few code comments.

Some more **general thoughts**:

I find myself not using logs that much outside of major add-ons, but I think that's primarily due to the boilerplate work required to make them useful. As a comparison point, I really like how VS Code handles this, with extensions being able to register their own output streams and that output being easily accessible to users. It also provides some other convenience features like being able to set the log level globally and per extension. Of course it's a different app with a different target audience, but I do think that some of these features would also benefit our add-on ecosystem

So this is to say, I do think there's merit in providing common logging facilities for add-ons.

However, I would also agree with @abdnh in that in order for this to be truly useful to add-on authors, we would need some **accompanying UI features** that lower the barrier of entry when asking users to debug issues.

Review Heatmap is the add-on where I've used logs the most in the past, and if I were to consider replacing the custom implementation there with built-in logging, I think the two most important features to me would be:

- a way for users to quickly access the debug logs, without needing to guide them through navigating to the right location, so similar to what @abdnh mentioned
- a way for users to temporarily increase the log level when digging deeper into an issue with them

I think that, with your current implementation, both of these actually don't seem to far off, and if you have time to work on them as part of this PR, I think that would make this a very nice overall bundle of improvements to add-on debugging.

Abdo mentioned adding a button to the add-on dialog, but since this would be an advanced path that users would not go down as commonly, IMHO I think it would be OK to hide the UI related to this away a bit more. The add-ons dialog already has quite a few buttons, so I think we have to keep an eye on not overloading it further.

The UX I'd have in mind then would be to use a context menu on the add-on list, which could look something like this:

| Add-on has logger | Add-on does not have logger |
| ----------- | ----------- |
| Screenshot_20240131_122420 | Screenshot_20240131_122506

The actions would only be enabled for add-ons that have a logger registered. Clicking on _Enable debugging_ would raise the debug level for the corresponding logger to `logging.DEBUG`. _View logs_ would open the logs folder for the add-on. (Copying the logs might also be nice to add, but we'd have to consider how to handle multiple sessions worth of logs and/or logs that exceed the clipboard size).

WDYT?

**Comment by glutanimate** [comment]
Regarding conflicts with existing custom logging implementations: I briefly tested Review Heatmap and Image Occlusion Enhanced and did not find any issues there.

File: qt/aqt/log.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+from typing import Any
+
+
+# this formatter instance is the same for all the handlers
+# (shortname is not a standard field, will get injected by record_factory)
+FORMATTER = logging.Formatter("%(asctime)s:%(shortname)s:%(name)s: %(message)s")
+
+
+
+def record_factory(old_factory: Any, *args, **kwargs) -> logging.LogRecord:
```
[the remark this answers is not in the document]
Comment (abdnh) [reply qt/aqt/log.py:28]: I agree.

File: qt/aqt/log.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        "WARN": "WRN",
+        "INFO": "INF",
+        "DEBUG": "DBG",
+    }.get(record.levelname, record.levelname)
+    return record
+
+
+class RotatingFileHandler(logging.handlers.RotatingFileHandler):
+    # The addon logs will be rotated once they reach MAXSIZE, to a total of BACKUPCOUNT
+    MAXSIZE = 3 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (abdnh) [reply qt/aqt/log.py:45]: The number is coming from my implementation, based on implementation in the AnkiHub add-on. The add-on makes heavy use of logs, and can actually fill up 15MB in a day. I agree another approach might be better for most add-ons.

File: qt/aqt/log.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/log.py]: got it, it's been removed now.

File: qt/aqt/addons.py [hunk 9]
```diff
@@ -665,6 +671,9 @@ def addonConfigHelp(self, module: str) -> str:
     def addonFromModule(self, module: str) -> str:
         return module.split(".")[0]
 
+    def getLogger(self, module: str) -> logging.Logger:
```
[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/addons.py:691]: I've kept around the `getLogger` as py3.9 doesn't have the get_logger (I'm not sure if 3.1* will have that get_logger at some point).

File: qt/aqt/log.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+            path.parent.mkdir(parents=True, exist_ok=True)
+
+            handler = RotatingFileHandler(path)
+            logger.addHandler(handler)
+            handler.setFormatter(FORMATTER)
+        return logger
+
+
+def config(path: Path | str, **kwargs) -> None:
+    """configure the mail logging
```
[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/log.py]: fixed now

File: qt/aqt/log.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/log.py]: I've switched to logging.handlers.TimedRotatingFileHandler base, with 1D rollover and 10 backups.

**Comment by cav71** [comment]
> Maybe another look over at how VS Code does things might be helpful here. The log structure there looks something as follows (simplified):
> 
> ```
> Code
> └── logs
>     └── 20240131T1400  (← session start time)
>         └── extensions
>             └── extension.a.log
> ```

Sounds good to me, I've added a pm.logFolder() to the ProfileManager:
```
~/Library/Application Support/Anki2
├── logs
│   └── addon
│       └── 566351439
│           └── 566351439.log
```
I'm not sure where anki stores the "session" id.

**Comment by cav71** [comment]
@glutanimate 
>I think that, with your current implementation, both of these actually don't seem to far off, and if you have time to work on them as part of this PR, I think that would make this a very nice overall bundle of improvements to add-on debugging.

Unfortunately I don't know which qt designer are you using (eg. I've noticed you have pyqt5/6), and I'm not familiar with the code structure to make these changes (I haven't used pyqt in a looong while).

But I've added three methods on the AddonManager class:
- addon_get_logger : returns a logging.Logger if addon module is present)
- addon_toggle_log_level : toggle loglevel to DEBUG on a module/addon (with ANKIDEV set to 1 you won't notice differences BTW)
- addon_get_logpath : return the log output Path

With these, it should be trivial to implement the UI callbacks you proposed.

As this code has grown considerably, what's the test strategy you have in place? I've noticed there's a qt/tests folder present.

File: qt/aqt/addons.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (abdnh) [reply qt/aqt/addons.py:697]: @cav71 this is about renaming AddonManager.getLogger(), not logging.getLogger()

[the remark this answers is not in the document]
Comment (cav71) [reply qt/aqt/addons.py]: done

**Comment by glutanimate** [comment]
Thanks @cav71, appreciate the follow-up and your laying down the groundwork for the UI changes. @dae if you are OK with the UI proposal above I can pick up the Qt work in a PR following this one.

On that note, I took the liberty to refactor the current implementation a bit in preparation of that. This is the [corresponding branch and diff](https://github.com/cav71/anki/compare/logging...glutanimate:anki:addon-logging-tweaks). I hope the commit log explains the changes, but my main goals were:

- To reduce the statefulness on `AddonManager` a little bit and simplify `LoggerManager` (e.g. by just using the assumed add-on log folder rather than checking the handlers for the current log file)
- To enable add-ons to more easily use logging without coupling themselves to a running Anki instance (e.g. when testing add-on code in unit tests)
- And finally some minor style changes and removing a few obsolete snippets now that the log storage location, etc. has changed

I would appreciate it if you could give the changes a look/try @cav71 and let me know if they are OK for you & functionally equivalent to what you had in mind. If so, we could maybe merge them into this PR. The checks should then also pass.

In terms of adding tests, coverage for `aqt` is rather low, and @dae will be able to better judge the necessity in this case, but if you would like to add some tests, I would imagine that would always be appreciated. It might actually be good to have some of the `Logger.manager` patching in `log.py`  tested, now that I think of it. Testing framework wise, we use `pytest` in `aqt/qt/tests`.

**Comment by cav71** [comment]
> Thanks @cav71, appreciate the follow-up and your laying down the groundwork for the UI changes. @dae if you are OK with the UI proposal above I can pick up the Qt work in a PR following this one.

Great thanks, that's a relief.


> On that note, I took the liberty to refactor the current implementation a bit in preparation of that. This is the [corresponding branch and diff](https://github.com/cav71/anki/compare/logging...glutanimate:anki:addon-logging-tweaks). I hope the commit log explains the changes, but my main goals were:

I think it is ready to go, I've made a single remark regarding `get_addon_logs_folder`: I think it would make sense using the new `def logFolder(self) -> str` from qt/aqt/profiles.py, same any other system folder. 

In general AddonManager seems too big of a class: it will be harder to test it in future.

Finally (as I don't have windows) I couldn't test the case when an addon is deleted/backed up.

> I would appreciate it if you could give the changes a look/try @cav71 and let me know if they are OK for you & functionally equivalent to what you had in mind. If so, we could maybe merge them into this PR. The checks should then also pass.

I'll merge `addon-logging-tweaks` as soon you're ok with it: the seems working fine.
 
> In terms of adding tests, coverage for `aqt` is rather low, and @dae will be able to better judge the necessity in this case, but if you would like to add some tests, I would imagine that would always be appreciated. It might actually be good to have some of the `Logger.manager` patching in `log.py` tested, now that I think of it. Testing framework wise, we use `pytest` in `aqt/qt/tests`.

Got it, thanks! Do you have the complete command line to run pytest (including the PYTHONPATHs) please?

**Comment by glutanimate** [comment]
Looks like I missed a couple of pylint checks. Should be fixed with ca8f7feaa82d90082d1ba665bfa62d6bec958f77 and dd59e33c6df7befcc5858714eefcf066297db8d2.

**Comment by cav71** [comment]
@glutanimate included these

> [ca8f7fe](https://github.com/ankitects/anki/commit/ca8f7feaa82d90082d1ba665bfa62d6bec958f77) and [dd59e33](https://github.com/ankitects/anki/commit/dd59e33c6df7befcc5858714eefcf066297db8d2).

File: qt/aqt/profiles.py [hunk 11]
```diff
@@ -309,6 +309,9 @@ def backupFolder(self) -> str:
     def collectionPath(self) -> str:
         return os.path.join(self.profileFolder(), "collection.anki2")
 
+    def log_folder(self) -> str:
+        return self._ensureExists(os.path.join(self.base, "logs"))
```
Comment (dae) [inline qt/aqt/profiles.py:313]: Is this only used by add-ons? If so, maybe something like addon_logs would make more sense?

**Comment by dae** [comment]
@abdnh @glutanimate could you give this post a thumbs up if you're happy to have this merged in now?

File: qt/aqt/profiles.py (hunk 11 above)
Comment (glutanimate) [reply qt/aqt/profiles.py:313]: Should we move `anki.log` into the folder perhaps? Happy to do so in a follow-up, or update the method name otherwise.

File: qt/aqt/log.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (abdnh) [reply qt/aqt/log.py]: Not a big deal, but some waitress logs are still shown. Eg:
```
2024-02-06 23:35:32,391:WARNING:waitress: unhandled close event
```

File: qt/aqt/profiles.py (hunk 11 above)
Comment (dae) [reply qt/aqt/profiles.py:313]: Most users will have an empty anki.log, due to the default log level. I don't think I've taken advantage of it once since it was first introduced, and would be just as happy to see it go, as there are very few issues that can't be better diagnosed with a traceback or a copy of the collection.

@cav71 would you mind renaming the method + folder name? Then I'll get this merged in.
